src/v/serde/envelope.h

```diff
@@ -67,77 +68,20 @@ struct checksum_envelope {
     static constexpr auto redpanda_serde_build_checksum = true;
 };
 
-namespace detail {
-
-template<typename T, typename = void>
-struct has_compat_attribute : std::false_type {};
-
-template<typename T>
-struct has_compat_attribute<
-  T,
-  std::void_t<decltype(std::declval<T>().redpanda_serde_compat_version)>>
-  : std::true_type {};
-
-template<typename T, typename = void>
-struct has_version_attribute : std::false_type {};
-
-template<typename T>
-struct has_version_attribute<
-  T,
-  std::void_t<decltype(std::declval<T>().redpanda_serde_version)>>
-  : std::true_type {};
-
-template<typename T, typename = void>
-struct inherits_from_envelope : std::false_type {};
-
-template<typename T>
-struct inherits_from_envelope<
-  T,
-  std::void_t<decltype(std::declval<T>().redpanda_inherits_from_envelope)>>
-  : std::true_type {};
-
 template<typename T>
-struct compat_version_has_serde_version_type {
-    static constexpr auto const value = std::is_same_v<
-      std::decay_t<decltype(std::declval<T>().redpanda_serde_compat_version)>,
-      version_t>;
+concept is_envelope = requires {
+    { T::redpanda_serde_version } -> std::convertible_to<serde::version_t>;
+    { T::redpanda_serde_compat_version } -> std::convertible_to<serde::version_t>;
 };
 
 template<typename T>
-struct version_has_serde_version_type {
-    static constexpr auto const value = std::is_same_v<
-      std::decay_t<decltype(std::declval<T>().redpanda_serde_version)>,
-      version_t>;
+concept is_checksum_envelope = is_envelope<T> && requires {
+    { T::redpanda_serde_build_checksum } -> std::convertible_to<bool>;
 };
 
-template<typename T, typename = void>
-struct has_checksum_attribute : std::false_type {};
-
 template<typename T>
-struct has_checksum_attribute<
-  T,
-  std::void_t<decltype(std::declval<T>().redpanda_serde_build_checksum)>>
-  : std::true_type {};
-
-} // namespace detail
-
-template<typename T>
-inline constexpr auto const is_envelope_v = std::conjunction_v<
-  detail::has_compat_attribute<T>,
-  detail::has_version_attribute<T>,
-  detail::compat_version_has_serde_version_type<T>,
-  detail::version_has_serde_version_type<T>>;
-
-template<typename T>
-inline constexpr auto const is_checksum_envelope_v = std::conjunction_v<
-  detail::has_compat_attribute<T>,
-  detail::has_version_attribute<T>,
-  detail::compat_version_has_serde_version_type<T>,
-  detail::version_has_serde_version_type<T>,
-  detail::has_checksum_attribute<T>>;
-
-template<typename T>
-inline constexpr auto const inherits_from_envelope_v
-  = detail::inherits_from_envelope<T>::value;
+concept inherits_from_envelope = requires {
+    { T::redpanda_inherits_from_envelope } -> std::convertible_to<bool>;
```

Comment:
This should probably be checked against `true`.

Reply:
yup. fixed and repushed.